pkg/publisher/combo/fanout.go

```diff
@@ -102,13 +116,49 @@ func (f *fanoutPublisher) PublishShardResult(
 		return p.PublishShardResult(ctx, shard, hostID, shardResultPath)
 	})
 
-	// Just return the first storage spec that we get
-	select {
-	case value := <-valueChannel:
-		return value, nil
-	case err := <-errorChannel:
-		return model.StorageSpec{}, err
+	type result struct {
+		Sender publisher.Publisher
+		Value  model.StorageSpec
 	}
+
+	timeoutChannel := make(chan bool, 1)
+	results := make([]result, 0)
+
+loop:
+	for {
+		select {
+		case value := <-valueChannel:
+			results = append(results, value)
+			if len(results) == len(f.publishers) {
+				// break because everyone returned
+				break loop
+			}
+
+			// start timeout for other results when first result is returned
+			if len(results) == 1 {
+				go func() {
+					// TODO: Make timeout value configurable
+					time.Sleep(time.Second * 2)
+					timeoutChannel <- true
+				}()
+			}
```

Comment:
If the result we got was from the preferred publisher, we don't need to start a timeout and can return immediately. Similarly, we can break out of the loop before everyone returned if we got a result from the preferred publisher because we know we are going to use it right away.

Reply:
Great feedback!